src/main/java/me/clip/placeholderapi/replacer/CharsReplacer.java

```diff
@@ -104,7 +108,7 @@ public String apply(@NotNull final String text, @Nullable final OfflinePlayer pl
         continue;
       }
 
-      final PlaceholderExpansion placeholder = lookup.apply(lowercaseIdentifierString);
+      final PlaceholderExpansion placeholder = lookup.apply(identifierString.toLowerCase(Locale.ROOT));
       if (placeholder == null) {
         builder.append(closure.head).append(identifierString);
 

```

Comment:
What Emily said, this only hurts readability and doesn't actually improve anything performance-wise.

Reply:
What about if the above condition `invalid` is true, the iteration continues to the next element and the `String#toLowerCase` was unnecessarily called before? Useless calls such as this should be avoided.

Reply:
Is the performance improvement really THAT significant?
If this is just a few ms, then changing this absolutely pointless and makes the code worse than it already is IMO.

Reply:
Yeah we're talking in literally like dozens of CPU cycles here, this changes absolutely nothing. Having the placeholders running on a separate thread would be a significantly bigger improvement than the micro optimization of this. Weighing the purported performance benefit vs the way it makes the code worse and harder to read, it is easy to make the call to leave it as is.

Reply:
If the `invalid` condition is true in all cases and this method is called more than once, I could say yes.